sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java

```diff
@@ -0,0 +1,186 @@
+        callbackManager.addErrorListener(errorListener, isDone && error != null, error);
+        return null;
+    }
+
+    @Override
+    public Request<T> withSessionListener(SessionListener sessionListener) {
+        throw new UnsupportedOperationException("AndroidAsyncHttp does not support SessionListeners");
+    }
+
+    protected AndroidAsyncHttpRequest(AsyncHttpClient client, Context context, String url, File file, Class<T> resultType) {
+        this.client = client;
+        this.context = context;
+        this.url = url;
+        this.file = file;
+        this.classOfResult = resultType;
+        this.callbackManager = new CallbackManager();
+        if (isSessionAboutToExpire()) {
+            Podio.client.forceRefreshTokens().withSessionListener(this);
+        } else {
+            performRequest(client, context, url, file);
+        }
+    }
+
+    private boolean isSessionAboutToExpire() {
+        return System.currentTimeMillis() > (Session.expires() - TEN_MINUTES);
+    }
+
+    @Override
+    public boolean onSessionChanged(String authToken, String refreshToken, String transferToken, long expires) {
+        performRequest(client, context, url, file);
+
+        return false;
+    }
+
+    public void performRequest(AsyncHttpClient client, Context context, String url, File file) {
+        try {
+            RequestParams params = new RequestParams();
+            params.put("source", file);
+            params.put("filename", file.getName());
+            client.post(context, url, getHeaders(), params, CONTENT_TYPE, new BaseJsonHttpResponseHandler<T>() {
+
+                @Override
+                protected T parseResponse(String rawJsonData, boolean isFailure) throws Throwable {
+                    T result;
+                    if (isFailure || classOfResult == null || classOfResult == Void.class) {
+                        result = null;
+                    } else {
+                        result = JsonParser.fromJson(rawJsonData, classOfResult);
+                    }
+                    return result;
+                }
+
+                @Override
+                public void onSuccess(int statusCode, Header[] headers, String rawJsonResponse, T response) {
+                    result = response;
+                    deliverResponse();
+                }
+
+                @Override
+                public void onFailure(int statusCode, Header[] headers, Throwable throwable, String rawJsonData, T errorResponse) {
+                    if (throwable instanceof ConnectTimeoutException || throwable instanceof SocketTimeoutException) {
+                        error = new NoResponseError(throwable);
+                    } else if (rawJsonData != null) {
+                        try {
+                            error = new ApiError(rawJsonData, statusCode, throwable);
+                        } catch (JsonSyntaxException jsonSyntaxException) {
+                            error = new PodioError(throwable);
```

Comment:
The difference is that if it's an ApiError then we have a json error from our api that we parse, otherwise we have NoResponseError and ConnectionError and for unknown exceptions we bake them in to the PodioError.